### PR TITLE
Add operation-based jitless wasm entry thunk

### PR DESCRIPTION
**Pull request by justinmichaud**
#### e7607367125042a0211b278c6a5cee43eb7fbdff
<pre>
Add operation-based jitless wasm entry thunk
<a href="https://bugs.webkit.org/show_bug.cgi?id=275397">https://bugs.webkit.org/show_bug.cgi?id=275397</a>

Reviewed by Yusuke Suzuki.

Previously, we tried to implement jitless wasm entry by interpreting
bytecode. It seems way easier to just call into cpp to prepare the frame
instead, and the performance difference should not be too big.

* JSTests/wasm/stress/cc-i32-kitchen-sink.js:
* JSTests/wasm/stress/cc-int-to-int-cross-module-with-exception.js:
* JSTests/wasm/stress/cc-int-to-int-cross-module.js:
* JSTests/wasm/stress/cc-int-to-int-jit-to-llint.js:
* JSTests/wasm/stress/cc-int-to-int-memory.js:
* JSTests/wasm/stress/cc-int-to-int-no-jit.js:
* JSTests/wasm/stress/cc-int-to-int-tail-call.js:
* JSTests/wasm/stress/cc-int-to-int.js:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::JSEntrypointInterpreterCallee::JSEntrypointInterpreterCallee):
(JSC::Wasm::JSEntrypointInterpreterCallee::calleeSaveRegistersImpl):
* Source/JavaScriptCore/wasm/WasmCallee.h:
(JSC::Wasm::jsEntrypointMetadataForGPR): Deleted.
(JSC::Wasm::jsEntrypointMetadataForFPR): Deleted.
(JSC::Wasm::dumpJSEntrypointInterpreterCalleeMetadata): Deleted.
* Source/JavaScriptCore/wasm/WasmLLIntPlan.cpp:
(JSC::Wasm::LLIntPlan::tryCreateInterpretedJSToWasmCallee):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:

Canonical link: <a href="https://commits.webkit.org/280485@main">https://commits.webkit.org/280485@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/82b55afe9baed014be25042406f96f0dd93085c3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56747 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36075 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9221 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60362 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7190 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43698 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7380 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45972 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5047 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58777 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33908 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48982 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26829 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30688 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6313 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6193 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/49839 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/52630 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6584 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62045 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/55988 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/660 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6686 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53232 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/663 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49037 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53250 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12555 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/568 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/77749 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31904 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12882 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32989 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34074 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32735 "Built successfully") | | |

### CODE REVIEW

**Comment by justinmichaud** [comment]
```
----------------------------------------------------------------------------------------------------------------------
|        subtest         |     ms       |     ms       |  b / a   | pValue (significance using False Discovery Rate) |
----------------------------------------------------------------------------------------------------------------------
| HashSet-wasm-Runtime   |403.769231    |396.561404    |0.982149  | 0.183358                                         |
| HashSet-wasm-Startup   |8.269231      |8.596491      |1.039576  | 0.220143                                         |
| gcc-loops-wasm-Runtime |2471.615385   |2438.561404   |0.986627  | 0.307595                                         |
| gcc-loops-wasm-Startup |5.826923      |5.964912      |1.023681  | 0.753967                                         |
| quicksort-wasm-Runtime |28.634615     |28.473684     |0.994380  | 0.656360                                         |
| quicksort-wasm-Startup |1.846154      |1.842105      |0.997807  | 0.961313                                         |
| richards-wasm-Runtime  |387.096154    |372.087719    |0.961228  | 0.112236                                         |
| richards-wasm-Startup  |2.365385      |2.035088      |0.860362  | 0.003977 (significant)                           |
| tsf-wasm-Runtime       |2605.134615   |2385.473684   |0.915682  | 0.103589                                         |
| tsf-wasm-Startup       |11.961538     |13.526316     |1.130817  | 0.000000 (significant)                           |
----------------------------------------------------------------------------------------------------------------------


a mean = 104.77569
b mean = 107.35013
pValue = 0.1531241713
(Bigger means are better.)
1.025 times better
Results ARE NOT significant
```

File: Source/JavaScriptCore/llint/WebAssembly.asm [hunk 1]
```diff
@@ -659,37 +628,56 @@ end
...
+        f(27)
+        f(28)
+        f(29)
+    end
 
-if ARM64 or ARM64E or ARMv7
-    const JSEntrypointInterpreterSpillSpaceStackAligned = (4 * SlotSize + StackAlignment - 1) & ~StackAlignmentMask
-else
-    const JSEntrypointInterpreterSpillSpaceStackAligned = (8 * SlotSize + StackAlignment - 1) & ~StackAlignmentMask
+if ASSERT_ENABLED or true
```
Comment (rkirsling) [inline Source/JavaScriptCore/llint/WebAssembly.asm]: Presumably the `or true` was temporary here (and four other times below)?

File: Source/JavaScriptCore/wasm/WasmOperations.cpp [hunk 2]
```diff
@@ -71,6 +71,133 @@ namespace WasmOperationsInternal {
 }
 #endif
 
+static constexpr unsigned gprToIndex(GPRReg r)
+{
+    for (unsigned i = 0; i < GPRInfo::numberOfArgumentRegisters; ++i) {
+        if (GPRInfo::toArgumentRegister(i) == r)
+            return i;
+    }
+    RELEASE_BASSERT_NOT_REACHED();
```
Comment (rkirsling) [inline Source/JavaScriptCore/wasm/WasmOperations.cpp]: Whoops, here and immediately below, presumably shouldn't be `BASSERT` if we're not in bmalloc.

File: Source/JavaScriptCore/wasm/WasmOperations.cpp [hunk 3]
```diff
@@ -71,6 +71,133 @@ namespace WasmOperationsInternal {
...
+            else
+                dst = gprToIndex(wasmFrameConvention.params[i].location.jsr().payloadGPR()) * sizeof(UCPURegister);
+            ASSERT(dst >= 0);
+
+            dataLogLnIf(WasmOperationsInternal::verbose, "* Register Arg ", i, " ", src, " -> ", dst);
+
+            if (type.isI32())
+                *access.operator()<uintptr_t>(registerSpace, dst) = *access.operator()<uint32_t>(entryFrame, src);
+            else if (type.isF32())
+                *access.operator()<uint64_t>(registerSpace, dst) = *access.operator()<uint32_t>(entryFrame, src);
```
Comment (rkirsling) [inline Source/JavaScriptCore/wasm/WasmOperations.cpp]: How come the left-hand type differs here? Just felt surprising to see.

File: Tools/Scripts/run-jsc-stress-tests [hunk 4]
```diff
@@ -2469,7 +2469,6 @@ def handleCollectionFile(collection)
                     
                     $benchmark = path.basename
                     $benchmarkDirectory = path.dirname
-
```
Comment (rkirsling) [inline Tools/Scripts/run-jsc-stress-tests]: I guess we should drop this line from the patch.

File: Source/JavaScriptCore/runtime/OptionsList.h [hunk 5]
```diff
@@ -579,7 +579,7 @@ bool hasCapacityToUseLargeGigacage();
     v(Bool, useWasmIPIntEpilogueOSR, true, Normal, "Allow IPInt to tier up during function epilogues"_s) \
     v(Bool, wasmIPIntTiersUpToBBQ, true, Normal, "Allow IPInt to tier up to BBQ"_s) \
     v(Bool, wasmIPIntTiersUpToOMG, true, Normal, "Allow IPInt to tier up to OMG"_s) \
-    v(Bool, useInterpretedJSEntryWrappers, false, Normal, "Allow some JS->Wasm wrappers to be replaced by jit-less versions."_s) \
+    v(Bool, useInterpretedJSEntryWrappers, true, Normal, "Allow some JS->Wasm wrappers to be replaced by jit-less versions."_s) \
```
Comment (Constellation) [inline Source/JavaScriptCore/runtime/OptionsList.h]: Let's not enable this until we fully support all cases in this operation. (It is fine to enable it for testing, but not enabling it right now when landing).

**Review by Constellation — COMMENTED** [review]
It seems that tests are failing. Can you take a look?

File: Source/JavaScriptCore/llint/WebAssembly.asm (hunk 1 above)
Comment (justinmichaud) [reply Source/JavaScriptCore/llint/WebAssembly.asm]: Yeah

File: Source/JavaScriptCore/wasm/WasmOperations.cpp (hunk 2 above)
Comment (justinmichaud) [reply Source/JavaScriptCore/wasm/WasmOperations.cpp]: Oh wow, oops!

File: Source/JavaScriptCore/runtime/OptionsList.h (hunk 5 above)
Comment (justinmichaud) [reply Source/JavaScriptCore/runtime/OptionsList.h]: Will do

File: Source/JavaScriptCore/wasm/WasmOperations.cpp (hunk 3 above)
Comment (justinmichaud) [reply Source/JavaScriptCore/wasm/WasmOperations.cpp]: We load the whole register. On armv7, we do a 32-bit gpr load and a 64-bit double load. On 64-bit, they are both loaded as 64-bit values

File: Source/JavaScriptCore/wasm/WasmCallee.h [hunk 6]
```diff
@@ -390,22 +198,31 @@ class JSEntrypointInterpreterCallee final : public JSEntrypointCallee,
 
     static constexpr ptrdiff_t offsetOfWasmCallee() { return OBJECT_OFFSETOF(JSEntrypointInterpreterCallee, wasmCallee); }
     static constexpr ptrdiff_t offsetOfWasmFunctionPrologue() { return OBJECT_OFFSETOF(JSEntrypointInterpreterCallee, wasmFunctionPrologue); }
-    static constexpr ptrdiff_t offsetOfMetadataStorage() { return offsetOfData(); }
+    static constexpr ptrdiff_t offsetOfFrameSize() { return OBJECT_OFFSETOF(JSEntrypointInterpreterCallee, frameSize); }
+
+    // Space for callee-saves; Not included in frameSize
+    static constexpr unsigned SpillStackSpaceAligned = WTF::roundUpToMultipleOf<stackAlignmentBytes()>(3 * sizeof(UCPURegister));
```
Comment (rkirsling) [inline Source/JavaScriptCore/wasm/WasmCallee.h:205]: This needs an `#include "StackAlignment.h"` to build.

File: Source/JavaScriptCore/wasm/WasmOperations.cpp [hunk 7]
```diff
@@ -71,6 +71,136 @@ namespace WasmOperationsInternal {
...
+            return i;
+    }
+    RELEASE_ASSERT_NOT_REACHED();
+    return 0;
+}
+
+JSC_DEFINE_NOEXCEPT_JIT_OPERATION(operationJSToWasmEntryWrapperBuildFrame, void, (void* sp, void* cfr))
+{
+    RELEASE_ASSERT(Options::useInterpretedJSEntryWrappers());
+    dataLogLnIf(WasmOperationsInternal::verbose, "operationJSToWasmEntryWrapperBuildFrame sp: ", RawPointer(sp), " fp: ", RawPointer(cfr));
```
Comment (rkirsling) [inline Source/JavaScriptCore/wasm/WasmOperations.cpp:95]: `WasmOperationsInternal::verbose` is currently guarded by `#if ENABLE(WEBASSEMBLY_OMGJIT)`.

**Review by Constellation — CHANGES_REQUESTED** [review]
Error message
```
microbenchmarks/wasm-cc-int-to-int.js.default-wasm: Could not open file: ../../JSTests/microbenchmarks/wasm-cc-int-to-int.wasm
```
is showing up in the JSTests results.

File: Source/JavaScriptCore/wasm/WasmOMGPlan.h [hunk 8]
```diff
@@ -67,6 +67,8 @@ class OMGPlan final : public Plan {
     bool m_completed { false };
     std::optional<bool> m_hasExceptionHandlers;
     uint32_t m_functionIndex;
+
```
Comment (Constellation) [inline Source/JavaScriptCore/wasm/WasmOMGPlan.h]: Remove this space line.

File: Source/JavaScriptCore/jit/FPRInfo.h [hunk 9]
```diff
@@ -155,7 +155,7 @@ class FPRInfo {
         return (unsigned)reg;
     }
 
-    static FPRReg toArgumentRegister(unsigned index)
+    static constexpr FPRReg toArgumentRegister(unsigned index)
     {
         ASSERT(index < numberOfArgumentRegisters);
```
Comment (Constellation) [inline Source/JavaScriptCore/jit/FPRInfo.h]: Use `ASSERT_UNDER_CONSTEXPR_CONTEXT`

File: Source/JavaScriptCore/jit/FPRInfo.h [hunk 10]
```diff
@@ -251,7 +251,7 @@ class FPRInfo {
         return result;
     }
 
-    static FPRReg toArgumentRegister(unsigned index)
+    static constexpr FPRReg toArgumentRegister(unsigned index)
     {
         ASSERT(index < 8);
```
Comment (Constellation) [inline Source/JavaScriptCore/jit/FPRInfo.h]: Use `ASSERT_UNDER_CONSTEXPR_CONTEXT`

File: Source/JavaScriptCore/jit/FPRInfo.h [hunk 11]
```diff
@@ -331,10 +331,10 @@ class FPRInfo {
         return registerForIndex[index];
     }
 
-    static FPRReg toArgumentRegister(unsigned index)
+    static constexpr FPRReg toArgumentRegister(unsigned index)
     {
         ASSERT(index < numberOfArgumentRegisters);
```
Comment (Constellation) [inline Source/JavaScriptCore/jit/FPRInfo.h]: Use `ASSERT_UNDER_CONSTEXPR_CONTEXT`

File: Source/JavaScriptCore/jit/GPRInfo.h [hunk 12]
```diff
@@ -405,10 +405,10 @@ class GPRInfo {
         return registerForIndex[index];
     }
     
-    static GPRReg toArgumentRegister(unsigned index)
+    static constexpr GPRReg toArgumentRegister(unsigned index)
     {
         ASSERT(index < numberOfArgumentRegisters);
```
Comment (Constellation) [inline Source/JavaScriptCore/jit/GPRInfo.h]: Use `ASSERT_UNDER_CONSTEXPR_CONTEXT`

File: Source/JavaScriptCore/jit/GPRInfo.h [hunk 13]
```diff
@@ -500,10 +500,10 @@ class GPRInfo {
         return registerForIndex[index];
     }
 
-    static GPRReg toArgumentRegister(unsigned index)
+    static constexpr GPRReg toArgumentRegister(unsigned index)
     {
         ASSERT(index < numberOfArgumentRegisters);
```
Comment (Constellation) [inline Source/JavaScriptCore/jit/GPRInfo.h]: Use `ASSERT_UNDER_CONSTEXPR_CONTEXT`

File: Source/JavaScriptCore/jit/GPRInfo.h [hunk 14]
```diff
@@ -631,7 +631,7 @@ class GPRInfo {
         return (unsigned)reg;
     }
 
-    static GPRReg toArgumentRegister(unsigned index)
+    static constexpr GPRReg toArgumentRegister(unsigned index)
     {
         ASSERT(index < numberOfArgumentRegisters);
```
Comment (Constellation) [inline Source/JavaScriptCore/jit/GPRInfo.h]: Use `ASSERT_UNDER_CONSTEXPR_CONTEXT`

File: Source/JavaScriptCore/jit/GPRInfo.h [hunk 15]
```diff
@@ -741,10 +741,10 @@ class GPRInfo {
         return registerForIndex[index];
     }
 
-    static GPRReg toArgumentRegister(unsigned index)
+    static constexpr GPRReg toArgumentRegister(unsigned index)
     {
         ASSERT(index < numberOfArgumentRegisters);
```
Comment (Constellation) [inline Source/JavaScriptCore/jit/GPRInfo.h]: Use `ASSERT_UNDER_CONSTEXPR_CONTEXT`

File: Source/JavaScriptCore/wasm/WasmOperations.cpp [hunk 16]
```diff
@@ -65,11 +65,139 @@ IGNORE_WARNINGS_BEGIN("frame-address")
...
 }
-#endif
+
+static constexpr unsigned gprToIndex(GPRReg r)
+{
+    for (unsigned i = 0; i < GPRInfo::numberOfArgumentRegisters; ++i) {
+        if (GPRInfo::toArgumentRegister(i) == r)
+            return i;
+    }
+    RELEASE_ASSERT_NOT_REACHED();
```
Comment (Constellation) [inline Source/JavaScriptCore/wasm/WasmOperations.cpp]: Use `RELEASE_ASSERT_NOT_REACHED_UNDER_CONSTEXPR_CONTEXT`.

File: Source/JavaScriptCore/wasm/WasmOperations.cpp [hunk 17]
```diff
@@ -65,11 +65,139 @@ IGNORE_WARNINGS_BEGIN("frame-address")
...
+    return 0;
+}
+
+static constexpr unsigned fprToIndex(FPRReg r)
+{
+    for (unsigned i = 0; i < FPRInfo::numberOfArgumentRegisters; ++i) {
+        if (FPRInfo::toArgumentRegister(i) == r)
+            return i;
+    }
+    RELEASE_ASSERT_NOT_REACHED();
```
Comment (Constellation) [inline Source/JavaScriptCore/wasm/WasmOperations.cpp]: Ditto.

File: Source/JavaScriptCore/wasm/WasmOperations.cpp [hunk 18]
```diff
@@ -65,11 +65,139 @@ IGNORE_WARNINGS_BEGIN("frame-address")
...
+    return 0;
+}
+
+JSC_DEFINE_NOEXCEPT_JIT_OPERATION(operationJSToWasmEntryWrapperBuildFrame, void, (void* sp, void* cfr))
```
Comment (Constellation) [inline Source/JavaScriptCore/wasm/WasmOperations.cpp]: Why not having `CallFrame*` for `cfr`?

File: Source/JavaScriptCore/wasm/WasmOperations.cpp [hunk 19]
```diff
@@ -65,11 +65,139 @@ IGNORE_WARNINGS_BEGIN("frame-address")
...
+                *access.operator()<uint64_t>(registerSpace, dst) = *access.operator()<uint32_t>(entryFrame, src);
+            else if (type.isI64() || type.isF64())
+                *access.operator()<uint64_t>(registerSpace, dst) = *access.operator()<uint64_t>(entryFrame, src);
+            else
+                RELEASE_ASSERT_NOT_REACHED();
+        }
+    }
+}
+
+JSC_DEFINE_NOEXCEPT_JIT_OPERATION(operationJSToWasmEntryWrapperBuildReturnFrame, void, (void* sp, void* cfr))
```
Comment (Constellation) [inline Source/JavaScriptCore/wasm/WasmOperations.cpp]: Ditto.

**Review by Constellation — APPROVED** [review]
r=me with comment.
Please ensure that this is still disabled when landing.

**Pull request merged**